### PR TITLE
Re-GET nodes during CIDR allocation (to avoid cascading bad resource version).

### PR DESCRIPTION
**Pull request by cjcullen**
Fixes the large-cluster CIDR allocation problem described in #20984.

### CODE REVIEW

**Comment by cjcullen** [comment]
@gmarek @alex-mohr

File: pkg/controller/node/nodecontroller.go [hunk 1]
```diff
@@ -514,14 +514,20 @@ func (nc *NodeController) reconcileNodeCIDRs(nodes *api.NodeList) {
...
-				nc.recordNodeStatusChange(&node, "CIDRNotAvailable")
+				nc.recordNodeStatusChange(n, "CIDRNotAvailable")
 				continue
 			}
-			glog.V(4).Infof("Assigning node %s CIDR %s", node.Name, podCIDR)
-			node.Spec.PodCIDR = podCIDR
-			if _, err := nc.kubeClient.Core().Nodes().Update(&node); err != nil {
+			glog.V(4).Infof("Assigning node %s CIDR %s", n.Name, podCIDR)
+			n.Spec.PodCIDR = podCIDR
+			if _, err := nc.kubeClient.Core().Nodes().Update(n); err != nil {
```
Comment (wojtek-t) [inline pkg/controller/node/nodecontroller.go]: @cjcullen - FYI you can solve this issue by just setting ResourceVersion to 0 just before the update.
If ResourveVersion is set to 0, this is treated as "unconditional update" and will be served correctly in pkg/registry/generic/etcd code:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/generic/etcd/etcd.go#L252

Comment (cjcullen) [inline pkg/controller/node/nodecontroller.go]: That feels scary. If there was a meaningful update in between my read and write, wouldn't I clobber it?

Comment (wojtek-t) [inline pkg/controller/node/nodecontroller.go]: Yes - it would.

**Pull request merged**